mypy/constraints.py

```diff
@@ -1018,10 +1018,22 @@ def visit_callable_type(self, template: CallableType) -> list[Constraint]:
             param_spec = template.param_spec()
 
             template_ret_type, cactual_ret_type = template.ret_type, cactual.ret_type
-            if template.type_guard is not None:
+            if template.type_guard is not None and cactual.type_guard is not None:
```

Comment:
Nice, thanks! I think this should fix a thing I ran into while testing https://github.com/python/mypy/pull/16939